### ISSUE TITLE
network.managed dns bug on Ubuntu

### ISSUE THREAD

**Issue by arsyee** [saltstack/salt#50703]
### Description of Issue/Question
network.managed sets dns-nameservers up with spaces included after each character
Resultint interfaces file:
```
root@jinjabug:~# cat /etc/network/interfaces
auto enp0s8
iface enp0s8 inet static
    address 192.168.1.70
    netmask 255.255.255.0
    dns-nameservers 8 . 8 . 8 . 8   8 . 8 . 4 . 4
auto enp0s3
iface enp0s3 inet dhcp
```

Note the spaces before and after the dots in dns-nameservers.
Any subsequent state.apply reports a change (i.e. state.apply recognizes the failure, tries to fix it, but sets up the dns-nameservers wrongly again, so during the next state.apply, it will run the exact same cycle):

```
          ID: enp0s8
    Function: network.managed
      Result: True
     Comment: Interface enp0s8 updated.
     Started: 12:28:32.366491
    Duration: 315.748 ms
     Changes:
              ----------
              interface:
                  ---
                  +++
                  @@ -2,5 +2,5 @@
                   iface enp0s8 inet static

                       address 192.168.1.70

                       netmask 255.255.255.0

                  -    dns-nameservers 8   .   8   .   8   .   8       8   .   8   .   4   .   4

                  +    dns-nameservers 8.8.8.8 8.8.4.4


              status:
                  Interface enp0s8 restart to validate

```

### Upstream tests  & Workaround
I tried to fix the bug on my own - it seems in debian_eth.jinja is responsible for joining the list of IP addresses. I tried to mess a bit with it without luck. I also saw that debian_eth.jinja had been refactored recently. I tried to patch the new jinja from upstream, but that also resulted in the same issue. It seemed like it's joining the string itself by the spaces, instead of the list of strings.

Unfortunately, jinja doesn't support list comprehension or other fancy Python features, so I couldn't really fix the problem there. Instead, I removed dns-nameservers from concat_opts, and changed debian_ip.py to join the nameservers in python before passing it to the Jinja template resolver:

```
1748,1749d1747
<     if 'inet' in opts[iface]['data'] and 'dns_nameservers' in opts[iface]['data']['inet']:
<         opts[iface]['data']['inet']['dns_nameservers'] = " ".join(opts[iface]['data']['inet']['dns_nameservers'])
```

This seems to work, but obviously not a long-term solution.

I tried to isolate the issue in Jinja and blame them, but without the whole SlatStack around, Jinja seemed to join the list of strings fine.

Also I'm a bit confused, because this use case seems so common and frequent that I doubt it can really be a failure, but this issue reproduced consistently on all my 3 VMs, and I could reproduce it on a 4th VM after a clean setup.

I'm happy to provide more information on the system, if needed. It's possible that I have a configuration issue, I just cannot see how.

### Setup
Clean Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-39-generic x86_64) system running in VirtualBox, only changes applied via Salt. See SLS files below.

### Steps to Reproduce Issue
Install Ubuntu (VM was cloned from clean installed Ubuntu with MAC addresses re-generated).
Install Salt with bootstrap.sh, change minion_id to 'jinjabug' (don't get confused, I still suspect this is a SaltStack issue), change minion config: master IP and master_finger provided, accept key on master. Apply states:

**top.sls**
```
base:
  'not master':
    - base.packages
    - base.configuration
  'jinjabug':
    - minions.jinjabug
```

**base/packages.sls**
```
mc:
  pkg.installed: []
ifupdown:
  pkg.installed: []
resolvconf:
  pkg.installed: []
netplan.io:
  pkg.purged: []
```

**base/configuration.sls**
```
cloud.cfg:
  file.managed:
    - name: /etc/cloud/cloud.cfg
    - source: salt://_res/cloud.cfg
sshkeys-tibi:
  ssh_auth.present:
    - user: <user> # removed
    - source: salt://_res/id_rsa_vm.pub
    - config: '%h/.ssh/authorized_keys'
sshkeys-root:
  ssh_auth.present:
    - user: root
    - source: salt://_res/id_rsa_vm.pub
    - config: '%h/.ssh/authorized_keys'
enp0s8_routes:
  network.routes:
    - name: enp0s8
    - routes:
      - name: default
        ipaddr: default
        netmask: 0.0.0.0
        gateway: 192.168.1.1
enp0s3_routes:
  network.routes:
    - name: enp0s3
    - routes: []
```

**jinjabug.sls**
```
'hostnamectl set-hostname jinjabug':
  cmd.run:
    - unless: test `cat /etc/hostname` = jinjabug
system:
  network.system:
    - enabled: True
    - hostname: jinjabug
    - nozeroconf: True
    - nisdomain: localdomain
    - require_reboot: True
enp0s8:
  network.managed:
    - enabled: True
    - type: eth
    - proto: none
    - ipaddr: 192.168.1.70
    - netmask: 255.255.255.0
    - dns:
      - 8.8.8.8
      - 8.8.4.4
enp0s3:
  network.managed:
    - enabled: True
    - type: eth
    - proto: dhcp

```

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

Below versions-report is from the master (it runs a raspbian). Minion Salt versions are the same, minions are VirtualBox VMs with Ubuntu.

```
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.2
      docker-py: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.0
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.9 (default, Mar  8 2015, 00:52:26)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: debian 8.0
         locale: UTF-8
        machine: armv7l
        release: 4.4.13-v7+
         system: Linux
        version: debian 8.0
```

### Notes

Ubuntu uses netplan.io out of the box, which seems to collide with /etc/network/interfaces configuration, so I had to remove the complete package and make a change in /etc/cloud/cloud.cfg to avoid getting a secondary dynamic IP address after setting up static configuration. I'm not sure if this is how Salt should work with Ubuntu. Also at some point ifup/ifdown didn't work, probably because netplan.io messed with the interfaces before. You should probably consider supporting netplan.io instead of the conventional /etc/network/interfaces configuration.

**Comment by yevi19** [comment]
I have the same issue on ubuntu 16.04 and salt 2018.3.3. Confirmed on multiple systems. I do not have the netplan problem described above as that is default for 18.04 not 16.04.

**/etc/network/interfaces**

    auto ens3
    iface ens3 inet static
        address 10.10.5.131
        netmask 255.255.252.0
        gateway 10.10.7.254
        dns-nameservers 1   0   .   1   0   .   7   .   2   5   4

**yaml**

    ens3:
      network.managed:
        - enabled: True
        - proto: static
        - type: eth
        - ipaddr: 10.10.5.13{{ identifier[1] }}
        - netmask: 255.255.252.0
        - gateway: 10.10.7.254
        - dns:
          - 10.10.7.254
        - require:
          - /etc/network/interfaces.d/50-cloud-init.cfg

**Salt Versions Master**

    Salt Version:
               Salt: 2018.3.3

    Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 2.4.2
          docker-py: Not Installed
              gitdb: 0.6.4
          gitpython: 1.0.1
              ioflo: Not Installed
             Jinja2: 2.8
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: 1.0.3
       msgpack-pure: Not Installed
     msgpack-python: 0.4.6
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.12 (default, Nov 12 2018, 14:36:49)
       python-gnupg: 0.3.8
             PyYAML: 3.11
              PyZMQ: 15.2.0
               RAET: Not Installed
              smmap: 0.9.0
            timelib: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.1.4

    System Versions:
               dist: Ubuntu 16.04 xenial
             locale: UTF-8
            machine: x86_64
            release: 4.13.0-1018-azure
             system: Linux
            version: Ubuntu 16.04 xenial

**Salt Version Minion**

    Salt Version:
               Salt: 2018.3.3

    Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 2.4.2
          docker-py: Not Installed
              gitdb: Not Installed
          gitpython: Not Installed
               ioflo: Not Installed
             Jinja2: 2.8
            libgit2: Not Installed
            libnacl: Not Installed
           M2Crypto: Not Installed
               Mako: 1.0.3
       msgpack-pure: Not Installed
     msgpack-python: 0.4.6
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 2.7.12 (default, Nov 12 2018, 14:36:49)
       python-gnupg: 0.3.8
             PyYAML: 3.11
              PyZMQ: 15.2.0
               RAET: Not Installed
              smmap: Not Installed
            timelib: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.1.4

    System Versions:
               dist: Ubuntu 16.04 xenial
             locale: UTF-8
            machine: x86_64
            release: 4.4.0-139-generic
             system: Linux
            version: Ubuntu 16.04 xenial

**Comment by yevi19** [comment]
I thought it might be related to https://github.com/saltstack/salt/pull/48844/

**Comment by max-arnold** [comment]
The bug is already fixed, just not released yet (hopefully it will be backported to stable branch): https://github.com/saltstack/salt/pull/50471#issuecomment-437836844

You can grab the fixed file from develop https://github.com/saltstack/salt/blob/develop/salt/modules/debian_ip.py#L382 and drop into your `_modules` dir. The actual fix is something like this:

https://github.com/saltstack/salt/commit/e4390e1f2ead670bdcba3d213d4aabac31951e2e

**Comment by yevi19** [comment]
Ok, I thought that was the case. Thank you.

**Comment by garethgreenaway** [comment]
@arsyee Thanks for the report.
@max-arnold I've marked the PR with the fix to be back-ported.  Thanks!

**Comment by yevi19** [comment]
I applied the fix https://github.com/saltstack/salt/commit/e4390e1f2ead670bdcba3d213d4aabac31951e2e to my environment and everything works as expected now. Thank you

**Comment by garethgreenaway** [comment]
Closing this out, if the problem persists please comment & we'll re-open the issue or feel free to open a new issue.

**Comment by timwsuqld** [comment]
@garethgreenaway Do we have an ETA as to when this'll be backported? Thanks

**Comment by garethgreenaway** [comment]
@timwsuqld It was backported to 2018.3 via https://github.com/saltstack/salt/pull/50804 and will be available in the next 2018.3 release as well as the upcoming Fluorine release.
